GroupShuffleSplit: Optionally partition by group size not just group

JochenSiegWork opened the pull request:
#### Reference Issues/PRs
Fixes [#9297](https://github.com/scikit-learn/scikit-learn/issues/9297)

#### What does this implement/fix?

This PR introduces an option to `GroupShuffleSplit` that generates the splits such that `train_size` and `test_size` can correspond to the number of samples. Previously `train_size` and `test_size` could only correspond to the number of groups. Splitting by the number of groups had the disadvantage that it was practically impossible to get a partition anything close to the desired train/test sizes when the group sizes were unbalanced.

Consider this toy example where the targeted train_size of 0.8 is actually 0.9 when the groups are very different in size:

```python
def get_shuffle_split(groups):
    X = y = np.ones(len(groups))
    gss = GroupShuffleSplit(n_splits=1, train_size=.8, random_state=0)
    train, test = next(gss.split(X, y, groups))
    return train, test
    
groups_balanced = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])  # balanced group sizes
groups_unbalanced = np.array([0, 0, 0, 0, 0, 1, 1, 1, 2, 4])  # unbalanced group sizes

train, test = get_shuffle_split(groups_balanced)
print(f"Balanced Train: index={train}, group={groups_balanced[train]}")
print(f"Balanced Test: index={test}, group={groups_balanced[test]}")
print(f"actual size Train: {len(train) / (len(train) + len(test))}")
# Balanced Train: index=[0 1 3 4 5 6 7 9], group=[0 1 3 4 5 6 7 9]
# Balanced Test: index=[2 8], group=[2 8]
# actual size Train: 0.8

train, test = get_shuffle_split(groups_unbalanced)
print(f"Unbalanced Train: index={train}, group={groups_unbalanced[train]}")
print(f"Unbalanced Test: index={test}, group={groups_unbalanced[test]}")
print(f"actual size Train: {len(train) / (len(train) + len(test))}")
# Unbalanced Train: index=[0 1 2 3 4 5 6 7 9], group=[0 0 0 0 0 1 1 1 4]
# Unbalanced Test: index=[8], group=[2]
# actual size Train: 0.9
```

Such imbalanced groups are common in molecular data used in drug discovery and molecular property prediction. When I clustered my molecular data and tried to use `GroupShuffleSplit` to generate train-test splits considering the groups from the clustering, I found that <50% of the train/test sizes deviated less than 5% from the target ratio. Sometimes considerably more. However, the group sizes I got can actually be grouped into typical 80/20 splits, but GroupShuffleSplit is just not doing it.

Therefore, I implemented a small function to enable GroupShuffleSplit to consider the group size. The algorithm greedily tries to find a heuristic solution to the bin-packing problem. It assigns the groups randomly until the group no longer fits into any bucket. Then, a group is assigned to the bucket whose size will be closer to the target size.

The interface I had in mind is using a `group_by` parameter and looks like the following. I am curious if this fits the sklearn API.

```python
gss = GroupShuffleSplit(n_splits=1, train_size=.8, group_by="size")
train, test = next(gss.split(X, y, groups_unbalanced))
```

I did a benchmark to analyze whether this feature is really helping. I used 13 data sets (mainly from [MoleculeNet](https://moleculenet.org/)) and clustered each of the 316 tasks with 4 clustering algorithms (scaffold, sphere exclusion, etc.) and then tried to generate train/test splits with `GroupShuffleSplit`. This makes 1264 train/test splits. For each split, I tried to call `GroupShuffleSplit` 10 times until it generated a split where neither the train nor test set deviates more than 5% from the target size.

Results:
- The current implementation of `GroupShuffleSplit` was successful in 44% (561 splits, average over 5 trials).
- The implementation proposed in this PR was successful in 99% (1247 splits, average over 5 trials).
- Both took about ~2 mins. My implementation took 111 sec on average and the current implementation took 131 sec. I made no effort to make the timings very comparable. Therefore, I would consider them comparably fast.

#### Any other comments?

I also tried an alternative implementation that works with a single permutation (just like `_iter_indices` of the `BaseShuffleSplit`). It also takes about 2 mins and gets 98% (1237 splits, average over 5 trials) right.

```python
    def _iter_indices_by_group_size(self, X, groups):
        # validation checks taken from super._iter_indices
        n_samples = _num_samples(X)
        n_train, n_test = _validate_shuffle_split(
            n_samples,
            self.test_size,
            self.train_size,
            default_test_size=self._default_test_size,
        )
        rng = check_random_state(self.random_state)

        classes, group_indices, group_counts = np.unique(
            groups,
            return_inverse=True,
            return_counts=True,
        )
        n_classes = len(classes)

        for i in range(self.n_splits):
            # random partition
            permutation = rng.permutation(n_classes)

            # fill the test set first
            test_sample_size = 0
            idx = 0
            for class_index in permutation:
                idx += 1
                test_sample_size += group_counts[class_index]
                if test_sample_size >= n_test:
                    break

            ind_test = permutation[:idx]
            ind_train = permutation[idx : (n_test + n_train)]

            train = np.flatnonzero(np.isin(group_indices, ind_train))
            test = np.flatnonzero(np.isin(group_indices, ind_test))

            yield train, test
```

Thanks for keeping sklearn maintained!

JochenSiegWork committed:
Partition by group size in GroupShuffleSplit

JochenSiegWork committed:
fix linting issues

JochenSiegWork committed:
fix code coverage

JochenSiegWork commented:
Hello,
I understand that managing a large open-source project such as this one can be extremely busy and time-consuming. However, some feedback would be nice. Is this feature interesting to the sklearn project at all? I am also happy to further contribute and refine the code/feature if you provide some feedback.

adrinjalali commented on sklearn/model_selection/_split.py:
```diff
@@ -2076,6 +2079,11 @@ class GroupShuffleSplit(GroupsConsumerMixin, BaseShuffleSplit):
         int, represents the absolute number of train groups. If None,
         the value is automatically set to the complement of the test size.
 
+    group_by : str, default="number"
```
`group_by` doesn't seem like the right name here. It has a specific meaning in data wrangling, and we're not doing an aggregation based on a group.

I'm not sure what would be better here. Maybe something like `split_size={"groups", "samples"}` would be better?

JochenSiegWork replied:
You're right, your suggestion more clearly names groups and samples as the basis for splitting. I renamed the parameter and the its values accordingly.

adrinjalali commented on sklearn/model_selection/_split.py:
```diff
@@ -2123,20 +2148,93 @@ def __init__(
             random_state=random_state,
         )
         self._default_test_size = 0.2
+        if group_by not in ["size", "number"]:
+            raise ValueError(
+                "Bad parameter 'group_by'. Allowed are 'size' and 'number'."
+            )
```
@jeremiedbb we don't use param validation in spliters?

JochenSiegWork replied:
Thanks for pointing this out. I am also not sure what the convention is but I am happy to move the code when someone can point me to the right location.

adrinjalali reviewed (commented):
I'm not sure how feasible it is to implement this, but right now it's broken:

```py
import numpy as np
from sklearn.model_selection import GroupShuffleSplit

X = np.ones(shape=(16, 2))
y = np.ones(shape=(16, 1))
groups = np.array([1, 2, 2, 3, 3, 3, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4])

gss = GroupShuffleSplit(n_splits=5, train_size=.7, random_state=42)

for train_index, test_index in gss.split(X, y, groups):
    print("indices: ", train_index, test_index)
    print("groups:  ", groups[train_index], groups[test_index])
```

produces two splits that are identical (the first two). In general, I'm not sure if it's feasible to solve this problem at all:

```
indices:  [0 3 4 5] [ 1  2  6  7  8  9 10 11 12 13 14 15]
groups:   [1 3 3 3] [2 2 4 4 4 4 4 4 4 4 4 4]
indices:  [0 3 4 5] [ 1  2  6  7  8  9 10 11 12 13 14 15]
groups:   [1 3 3 3] [2 2 4 4 4 4 4 4 4 4 4 4]
indices:  [1 2 3 4 5] [ 0  6  7  8  9 10 11 12 13 14 15]
groups:   [2 2 3 3 3] [1 4 4 4 4 4 4 4 4 4 4]
indices:  [ 3  4  5  6  7  8  9 10 11 12 13 14 15] [0 1 2]
groups:   [3 3 3 4 4 4 4 4 4 4 4 4 4] [1 2 2]
indices:  [ 0  6  7  8  9 10 11 12 13 14 15] [1 2 3 4 5]
groups:   [1 4 4 4 4 4 4 4 4 4 4] [2 2 3 3 3]
```

JochenSiegWork committed:
Merge branch 'main' into group_shuffle_split_by_size

JochenSiegWork committed:
rename group_by parameter to split_size

JochenSiegWork commented:
> I'm not sure how feasible it is to implement this, but right now it's broken:
> 
> ```python
> import numpy as np
> from sklearn.model_selection import GroupShuffleSplit
> 
> X = np.ones(shape=(16, 2))
> y = np.ones(shape=(16, 1))
> groups = np.array([1, 2, 2, 3, 3, 3, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4])
> 
> gss = GroupShuffleSplit(n_splits=5, train_size=.7, random_state=42)
> 
> for train_index, test_index in gss.split(X, y, groups):
>     print("indices: ", train_index, test_index)
>     print("groups:  ", groups[train_index], groups[test_index])
> ```
> 
> produces two splits that are identical (the first two). In general, I'm not sure if it's feasible to solve this problem at all:
> 
> ```
> indices:  [0 3 4 5] [ 1  2  6  7  8  9 10 11 12 13 14 15]
> groups:   [1 3 3 3] [2 2 4 4 4 4 4 4 4 4 4 4]
> indices:  [0 3 4 5] [ 1  2  6  7  8  9 10 11 12 13 14 15]
> groups:   [1 3 3 3] [2 2 4 4 4 4 4 4 4 4 4 4]
> indices:  [1 2 3 4 5] [ 0  6  7  8  9 10 11 12 13 14 15]
> groups:   [2 2 3 3 3] [1 4 4 4 4 4 4 4 4 4 4]
> indices:  [ 3  4  5  6  7  8  9 10 11 12 13 14 15] [0 1 2]
> groups:   [3 3 3 4 4 4 4 4 4 4 4 4 4] [1 2 2]
> indices:  [ 0  6  7  8  9 10 11 12 13 14 15] [1 2 3 4 5]
> groups:   [1 4 4 4 4 4 4 4 4 4 4] [2 2 3 3 3]
> ```

Yes, the current code is it not working sufficiently for many datasets. However, the aim of this PR is slightly different than the problem with identical splits you show in your example. I am also quite sure that solving identical splits is not completely possible, since some datasets are just too small, have only a few groups or have highly imbalanced group sizes, constraining the splitting procedure too much.

The problem addressed in this PR is that the actual size (in number of samples) of train and test set is very different to `train_size` and `test_size` parameters when the group sizes are imbalanced. This can be problematic and producing tiny train sets and hugh test sets in applications where imbalanced groups are frequent, like in drug discovery, cheminformatics and bioinformatics. In addition, other splitters like GroupKFold also split by the number of samples not the number of groups. This PR would help to align the functionality of the different splitters.

When I use `split_size='samples'` in your example the train/test sizes are now better aligned with 70% samples in the training set. However, the problem with the identical split persists and in this example there are now even 3 identical splits, which is not surprising considering the `split_size='samples'` adds another constrain to the splitting. I think small example datasets like this might not be very representative here, since using only 4 groups means there are only 2^4 possible assignments to train/test set, which is further constrained by the train/test_size parameters. I probably could provide some analysis on some real datasets if this is helpful.

I hope this helps to answer some questions. Let me know what I can do to improve the PR.


```python
import numpy as np
from sklearn.model_selection import GroupShuffleSplit


X = np.ones(shape=(16, 2))
y = np.ones(shape=(16, 1))
groups = np.array([1, 2, 2, 3, 3, 3, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4])

gss = GroupShuffleSplit(n_splits=5, train_size=.7, random_state=42, split_size="groups")

print("split_size='groups':")
for train_index, test_index in gss.split(X, y, groups):
    print("indices: ", train_index, test_index)
    print("groups:  ", groups[train_index], groups[test_index])

print()
print("split_size='samples'")
gss = GroupShuffleSplit(n_splits=5, train_size=.7, random_state=42, split_size="samples")

for train_index, test_index in gss.split(X, y, groups):
    print("indices: ", train_index, test_index)
    print("groups:  ", groups[train_index], groups[test_index])
```


```
split_size='groups':
indices:  [0 3 4 5] [ 1  2  6  7  8  9 10 11 12 13 14 15]
groups:   [1 3 3 3] [2 2 4 4 4 4 4 4 4 4 4 4]
indices:  [0 3 4 5] [ 1  2  6  7  8  9 10 11 12 13 14 15]
groups:   [1 3 3 3] [2 2 4 4 4 4 4 4 4 4 4 4]
indices:  [1 2 3 4 5] [ 0  6  7  8  9 10 11 12 13 14 15]
groups:   [2 2 3 3 3] [1 4 4 4 4 4 4 4 4 4 4]
indices:  [ 3  4  5  6  7  8  9 10 11 12 13 14 15] [0 1 2]
groups:   [3 3 3 4 4 4 4 4 4 4 4 4 4] [1 2 2]
indices:  [ 0  6  7  8  9 10 11 12 13 14 15] [1 2 3 4 5]
groups:   [1 4 4 4 4 4 4 4 4 4 4] [2 2 3 3 3]

split_size='samples'
indices:  [ 1  2 12  9  7  8 11 14 15  6 13 10] [5 4 3 0]
groups:   [2 2 4 4 4 4 4 4 4 4 4 4] [3 3 3 1]
indices:  [12  9 15  8  6 11  7  0 13 10 14] [1 2 5 4 3]
groups:   [4 4 4 4 4 4 4 1 4 4 4] [2 2 3 3 3]
indices:  [ 7 10 14  9 15 12  0  8  6 11 13] [1 4 3 2 5]
groups:   [4 4 4 4 4 4 1 4 4 4 4] [2 3 3 2 3]
indices:  [ 7 15  9 11 10 13 14 12  6  8  0] [1 3 4 5 2]
groups:   [4 4 4 4 4 4 4 4 4 4 1] [2 3 3 3 2]
indices:  [ 5  3  8  6 15  4 13 12  7 10 11 14  9] [1 2 0]
groups:   [3 3 4 4 4 3 4 4 4 4 4 4 4] [2 2 1]
```
